packages/bvaughn-architecture-demo/pages/index.tsx

```diff
@@ -0,0 +1,129 @@
+  const [deferredFilterByText, setDeferredFilterByText] = useState<string>("");
+
+  // Changing the focus range may cause us to suspend (while fetching new info from the backend).
+  // Wrapping it in a transition enables us to show the older set of messages (in a pending state) while new data loads.
+  // This is less jarring than the alternative of unmounting all messages and rendering a fallback loader.
+  const [range, setRange] = useState<Range | null>(null);
+  const [deferredRange, setDeferredRange] = useState<Range | null>(null);
+
+  // Using a deferred values enables the focus UI to update quickly,
+  // and the slower operation of Suspending to load points to be deferred.
+  //
+  // It also allows us to update the UI slightly, before we suspend to fetch new data,
+  // to indicate that what's currently being showed is stale.
+  const [isTransitionPending, startTransition] = useTransition();
+
+  const debouncedSetDeferredRange = useDebouncedCallback((newRange: Range | null) => {
+    startTransition(() => {
+      setDeferredRange(newRange);
+    });
+  }, FOCUS_DEBOUNCE_DURATION);
+
+  const updateFocusRange = useCallback(
+    (newRange: Range | null, debounce: boolean) => {
+      setRange(newRange);
+
+      // Focus values may change rapidly (e.g. during a click-and-drag)
+      // In this case, React's default high/low priority split is helpful, but we can do more.
+      // Debouncing a little before starting the transition helps avoid sending a lot of unused requests to the backend.
+      if (debounce) {
+        debouncedSetDeferredRange(newRange);
+      } else {
+        startTransition(() => {
+          setDeferredRange(newRange);
+        });
+      }
+    },
+    [debouncedSetDeferredRange]
+  );
+
+  const focusContext = useMemo<FocusContextType>(
+    () => ({
+      isTransitionPending,
+      rangeForDisplay: range,
+      range: deferredRange,
+      update: updateFocusRange,
+    }),
+    [deferredRange, isTransitionPending, range, updateFocusRange]
+  );
+
+  const updateFilters = useCallback((newFilterByText: string, newLevelFlags: ConsoleLevelFlags) => {
+    setLevelFlags(newLevelFlags);
+    setFilterByText(newFilterByText);
+    startTransition(() => {
+      setDeferredFilterByText(newFilterByText);
+    });
+  }, []);
+
+  const consoleFiltersContext = useMemo<ConsoleFiltersContextType>(
+    () => ({
+      filterByDisplayText: filterByText,
+      filterByText: deferredFilterByText,
+      isTransitionPending,
+      levelFlags,
+      update: updateFilters,
+    }),
+    [deferredFilterByText, filterByText, isTransitionPending, levelFlags, updateFilters]
+  );
+
+  // TODO Once we have a client implementation to interface with Replay backend,
+  // the app can inject a wrapper one that also reports cache hits and misses to this UI in a debug panel.
+
+  return (
+    <FocusContext.Provider value={focusContext}>
+      <ConsoleFiltersContext.Provider value={consoleFiltersContext}>
```

Comment:
This component uses React primitives (like `useState` and context) to setup and share "global" state with other parts of the app. This gives us a nice point to integrate with other React APIs (like [transitions](https://reactjs.org/docs/hooks-reference.html#usetransition)).